tests/integration/suite/daprd/workflow/crossapp/appdown.go

```diff
@@ -0,0 +1,165 @@
+	a.registry1 = task.NewTaskRegistry()
+	a.registry2 = task.NewTaskRegistry()
+
+	appID1 := uuid.New().String()
+	appID2 := uuid.New().String()
+
+	// App2: Activity that will be called before the app goes down
+	a.registry2.AddActivityN("ProcessData", func(ctx task.ActivityContext) (any, error) {
+		var input string
+		if err := ctx.GetInput(&input); err != nil {
+			return nil, fmt.Errorf("failed to get input in app2: %w", err)
+		}
+
+		// Sleep for a while to ensure the workflow is actively executing when app2 is stopped
+		time.Sleep(10 * time.Second)
+
+		return fmt.Sprintf("Processed by app2: %s", input), nil
+	})
+
+	// App1: Orchestrator that calls app2
+	a.registry1.AddOrchestratorN("AppDownWorkflow", func(ctx *task.OrchestrationContext) (any, error) {
+		var input string
+		if err := ctx.GetInput(&input); err != nil {
+			return nil, fmt.Errorf("failed to get input in orchestrator: %w", err)
+		}
+
+		// Call app2 activity
+		var result string
+		err := ctx.CallActivity("ProcessData",
+			task.WithActivityInput(input),
+			task.WithAppID(appID2)).
+			Await(&result)
+		if err != nil {
+			// Return error message to verify it's handled properly
+			return fmt.Sprintf("Error occurred: %v", err), nil
+		}
+
+		return result, nil
+	})
+
+	a.daprd1 = daprd.New(t,
+		daprd.WithInMemoryActorStateStore("mystore"),
+		daprd.WithPlacementAddresses(a.place.Address()),
+		daprd.WithScheduler(a.sched),
+		daprd.WithAppID(appID1),
+		daprd.WithAppPort(app1.Port()),
+		daprd.WithLogLevel("debug"),
+	)
+	a.daprd2 = daprd.New(t,
+		daprd.WithInMemoryActorStateStore("mystore"),
+		daprd.WithPlacementAddresses(a.place.Address()),
+		daprd.WithScheduler(a.sched),
+		daprd.WithAppID(appID2),
+		daprd.WithAppPort(app2.Port()),
+		daprd.WithLogLevel("debug"),
+	)
+
+	return []framework.Option{
+		framework.WithProcesses(a.place, a.sched, db, app1, app2, a.daprd1, a.daprd2),
+	}
+}
+
+func (a *appdown) Run(t *testing.T, ctx context.Context) {
+	a.sched.WaitUntilRunning(t, ctx)
+	a.place.WaitUntilRunning(t, ctx)
+	a.daprd1.WaitUntilRunning(t, ctx)
+	a.daprd2.WaitUntilRunning(t, ctx)
+	t.Cleanup(func() {
+		a.daprd2.Cleanup(t)
+	})
+
+	wctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	// Start workflow listeners for each app
+	client1 := client.NewTaskHubGrpcClient(a.daprd1.GRPCConn(t, ctx), backend.DefaultLogger())
+	client2 := client.NewTaskHubGrpcClient(a.daprd2.GRPCConn(t, wctx), backend.DefaultLogger())
+
+	// Start listeners for each app
+	err := client1.StartWorkItemListener(ctx, a.registry1)
+	assert.NoError(t, err)
+	err = client2.StartWorkItemListener(wctx, a.registry2)
+	assert.NoError(t, err)
+
+	id, err := client1.ScheduleNewOrchestration(ctx, "AppDownWorkflow", api.WithInput("Hello from app1"))
+	assert.NoError(t, err)
+
+	// Wait a bit for the workflow to start executing & Stop app2 to simulate app going down
+	time.Sleep(1 * time.Second)
+	cancel()
+	a.daprd2.Cleanup(t)
+
+	// Expect completion to hang, so timeout
```

Comment:
Why do we expect completion to hang?

Reply:
The workflow engine is designed to retry failed activities indefinitely until they succeed. In a real-world scenario, if an app goes down and never comes back up, the workflow would hang forever. However, since we can't let tests run indefinitely, I use a context timeout to verify that the workflow does indeed hang (as expected) rather than completing successfully. I've tested the transient app going down and back up again, via the restart test.